core/logger/zap_disk_logging.go

```diff
@@ -184,3 +85,56 @@ func (l *zapDiskLogger) pollDiskSpace() {
 		}
 	}
 }
+
+func newRotatingFileLogger(zcfg zap.Config, c Config, cores ...zapcore.Core) (*zapDiskLogger, func() error, error) {
+	defaultCore, defaultCloseFn, err := newDefaultLoggingCore(zcfg, c.UnixTS)
+	if err != nil {
+		return nil, nil, err
+	}
+	cores = append(cores, defaultCore)
+
+	diskLogLevel := zap.NewAtomicLevelAt(zapcore.DebugLevel)
+	diskCore, diskErr := newDiskCore(diskLogLevel, c)
+	if diskErr != nil {
+		defaultCloseFn()
+		return nil, nil, diskErr
+	}
+	cores = append(cores, diskCore)
+
+	core := zapcore.NewTee(cores...)
+	l, diskCloseFn, err := newLoggerForCore(zcfg, core)
+	if err != nil {
+		defaultCloseFn()
+		return nil, nil, err
+	}
+
+	lggr := &zapDiskLogger{
+		config: c,
+
+		pollDiskSpaceStop: make(chan struct{}),
+		pollDiskSpaceDone: make(chan struct{}),
+		zapLogger:         *l,
+		diskLogLevel:      diskLogLevel,
+	}
+
+	go lggr.pollDiskSpace()
+
+	var once sync.Once
+	closeLogger := func() error {
+		once.Do(func() {
+			close(lggr.pollDiskSpaceStop)
+			<-lggr.pollDiskSpaceDone
+		})
+
+		innerErr := lggr.Sync()
+		if err != nil {
+			return innerErr
```

Comment:
should the other close funcs be called even if there is an error? won't those resources be stranded forever?

and why is once.Do wrapping only part of the close'ings? i would have expected it to wrap all of the all of them.

Reply:
Good catch! Yes they should, and I'll also move them into the `once.Do` block as well.